[inductor max autotune] Detailed autotuning result logs ( machine-readable )

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119701
* #119600
* #119009
* #119008
* #119601
* #119007
* #119598
* #119597
* #119006
* __->__ #119004
* #120620

This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames

kadeng committed:
[inductor max autotune] Detailed autotuning result logs ( machine-readable )

This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented on torch/_inductor/debug.py:
```diff
@@ -481,6 +488,65 @@ def draw_orig_fx_graph(self, gm: torch.fx.GraphModule, nodes: SchedulerNodeList)
+            except Exception as e:
+                pass
+            try:
+                node_info["stride"] = str(node.get_stride())
```
Did that.

kadeng commented on torch/_inductor/debug.py:
```diff
@@ -481,6 +488,65 @@ def draw_orig_fx_graph(self, gm: torch.fx.GraphModule, nodes: SchedulerNodeList)
+                "type": type(node).__name__,
+            }
+            try:
+                node_info["layout"] = str(node.get_layout())
```
Here I left it as is, it should always be possible to get a string representation of the layout, even if it contains symbolic shapes. And that's ok in this case.

jansel commented on torch/_inductor/util_autotuning_log_parser.py:
```diff
@@ -0,0 +1,244 @@
+            }
+        else:
+            res = {
+                "size": (int(vals[0]), int(vals[1]), int(vals[3])),
```
Won't the casts to ints here fail with dynamic shapes?

kadeng replied:
Yes. Using the parser on dynamic shapes would be pointless, though. The point of this utility class (as I see it) is to enable microbenchmark analysis where we have obtained timings for specific fixed shapes. While this could be made to gracefully reject entries with dynamic shapes, or we could try to convert all dynamic shapes to specific shapes by using shape hints, I don't think it's worth to invest the time to develop and test that change. The primary usecase for this is when doing experiments with dynamic_shapes=False. I will add a comment that clarifies this. Please tell me if you see a usecase for dynamic_shapes=True, though. We could also add dynamic shape support in a separate PR if we see a need.

jansel commented on torch/_inductor/util_autotuning_log_parser.py:
```diff
@@ -0,0 +1,244 @@
+    @classmethod
+    def parse_layout(cls, layout_str):
+        match: Optional[re.Match[str]] = re.search(
+            r"size=\[([0-9]+), ([0-9]+)(, ([0-9]+))?\], stride=\[([0-9]+), ([0-9]+)(, ([0-9]+))?\]",
```
If we leave the symbolic names (s0) in the size, this regexp will fail for dynamic shapes.

kadeng replied:
True, see above.

jansel reviewed (changes requested):
Add a test that checks this works with dynamic shapes.

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/util_autotuning_log_parser.py:
On further thought, can we just use `json.dump`/`json.load` and remove the need for bespoke log parser entirely?

kadeng replied:
I don't see a good way how to do that. If we remove the parser, it will be very hard to make sense of the logs. It's not entirely trivial to interpret and analyze the raw log entries correctly without it. So we would have to add docs for the log format that are invariably going to be more complex and harder to understand than this code. 

If I changed the logging logic to make it trivial to interpret, we would add processing time to the logger - which I would consider a no-go, logging should have low overhead . And I would have to omit details, which means we would lose much of the ability to retroactively extract more info out of raw logs. The raw logs contain a lot of details that would help debug performance issues, which are unneccessary in an aggregate view. 

This Autotuning logger and log parser was developed with a very specific usecase in mind: Enabling the Cutlass backend Microbenchmarks and analyzing GEMM performance on the models I tested it on. And I think it's useful, so I created this PR to keep it. It's a performance debugging tool - machine readable and analyzable performance debug logs.

So, counter question: Do you see usecases beyond what I described? I would not recommend, for example, to use this for production logging or such.

jansel replied:
You would replace the parser with `json.load`.  Basically rather than writing own parser, you use a standard one.

kadeng replied:
I'm not sure I see what you mean. If you mean parsing the entire log file with a single call to json.load, this does not work for the following reasons:

 * The file is appended to line by line. If we would records as a json list, we would require closing and opening brackets, which requires to know when the records start and when they end, contradicting continuous logging.
 * The parser does a lot more than just reading. It loads the raw records ( JSON strings each ). Then, depending on the granularity that you want to look at it ( least granular: raw records, most granular: aggregated analysis ) it processes that log. So you can get raw records with backend specifics like Cutlass operator type to records, over general records which represent the lowest common denominator between backends ( e.g. M, N, K, Grid sizes, timings, layout info etc. ) to aggregate analysis (Aggregate by problem size and backend type, find fastest op config for each group etc..). This cannot be replaced by a single call to json.load.

jansel replied:
Each log entry would be one json blob.

kadeng replied:
Sure, we can do that. This logging is not critical in any case, it's a performance debugging / microbenchmarking tool. So we can still freely decide if, and if yes, where we would want to place it. Looking at the folder structure below benchmarks, I think something below benchmarks/operator_benchmark/ would be a good location.

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on third_party/kineto:
Revert this change

kadeng replied:
Sorry, overlooked your comment here. That submodule got accidentally locked in during a rebase I guess. Removed it.

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel commented on third_party/kineto:
Submodule changes leaked into this PR again

jansel commented on third_party/kineto:
Still not fixed

jansel commented:
You want to do something like:
```
git checkout origin/main third_party/kineto
git submodule update
```

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel commented on third_party/kineto:
Still not fixed

kadeng committed:
Update on "[inductor max autotune] Detailed autotuning result logs ( machine-readable )"


This diff introduces a new separate logging of autotuning results,
with the intention of making the results analyzable, specifically
those for the new experimental Cutlass backend.

Results are logged as text files with one JSON document corresponding to a single benchmark result per line.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]
